File: src/main/java/no/mesan/fag/patterns/timesheet/facade/PoiAdapter.java
package no.mesan.fag.patterns.timesheet.facade;

import java.util.HashMap;
import java.util.Map;

import no.mesan.fag.patterns.timesheet.format.StyleFactory.StyleName;
import no.mesan.fag.patterns.timesheet.format.StyleSpec;
import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.CellStyle;
import org.apache.poi.ss.usermodel.FormulaEvaluator;
import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.usermodel.Workbook;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

/** Adapter vår fasade til POI Workbook. */
public class PoiAdapter {
    /// HINT Dette kan være start på en fasadeklasse - du må nok inn med mer
    private final Workbook workbook = new XSSFWorkbook();
    private final Sheet sheet;
    private final Map<StyleName, CellStyle> styles = new HashMap<>();

    public PoiAdapter(final String title, final Map<StyleName, StyleSpec> map) {
        sheet= workbook.createSheet(title);
        createWbStyles(map);
        printSetup();
    }

    /// HINT Skygger av løsning...
    private void printSetup() { }
    public void addData(final SpreadSheet values) {}
    public Workbook create() {return null;}

    /**
     * Konverter StyleSpecs til CellStyles.
     * @param map Map av navn til StyleSpec
     */
    private void createWbStyles(final Map<StyleName, StyleSpec> map){
        for (final Map.Entry<StyleName, StyleSpec> entry : map.entrySet()) {
            styles.put(entry.getKey(), entry.getValue().createStyle(workbook));
        }
    }

    /**
     * Evaluer alle formler på et gitt ark på nytt.
     * @return Antall kolonner
     */
    private int reevaluateSheet() {
        int maxCols= 0;
        final FormulaEvaluator evaluator = sheet.getWorkbook().getCreationHelper().createFormulaEvaluator();
        for (final Row r : sheet) {
            int cols= 0;
            for (final Cell c : r) {
                ++cols;
                if (c.getCellType() == Cell.CELL_TYPE_FORMULA) {
                    evaluator.evaluateFormulaCell(c);
                }
            }
            maxCols = Math.max(maxCols, cols);
        }
        return maxCols;
    }
}

File: src/main/scala/no/mesan/fag/patterns/scala/timesheet/facade/PoiAdapter.scala
package no.mesan.fag.patterns.scala.timesheet.facade

import no.mesan.fag.patterns.scala.timesheet.format.{Styles, StyleName}
import no.mesan.fag.patterns.scala.timesheet.data.ValueMatrix

import org.apache.poi.ss.usermodel._
import org.apache.poi.xssf.usermodel.XSSFWorkbook

import java.io.FileOutputStream

import scala.collection.JavaConversions._


/** Adapter vår fasade til POI Workbook. */
class PoiAdapter(title: String, styleMap: Map[StyleName, Styles]) {
  private val workbook = new XSSFWorkbook
  private val sheet = workbook.createSheet(title)
  private val styles: Map[StyleName, CellStyle] = ??? /// HINT styleMap kan mappes vha createStyle

  printSetup

  /** Sett utskriftsoppsett. */
  private def printSetup {
    val printSetup: PrintSetup = sheet.getPrintSetup
    printSetup.setLandscape(true)
    sheet.setFitToPage(true)
    sheet.setHorizontallyCenter(true)
  }

  def create: Workbook = ??? /// HINT Fyll ut

  /** Evaluer alle formler på et gitt ark på nytt. */
  private def reevaluateSheet: Int = {
    var maxCols = 0
    val evaluator: FormulaEvaluator = sheet.getWorkbook.getCreationHelper.createFormulaEvaluator
    for (r <- sheet.rowIterator()) {
      var cols = 0
      for (c <- r.cellIterator()) {
        cols += 1
        if (c.getCellType == Cell.CELL_TYPE_FORMULA)  evaluator.evaluateFormulaCell(c)
      }
      maxCols = Math.max(maxCols, cols)
    }
    maxCols
  }

  /** Autojuster alle kolonner på et ark. */
  private def autoSizeSheet(noOfColumns: Int) {} /// HINT Fyll ut

  /** Legg til dataene. */
  def addData(values: SpreadSheet) = ???

  private def createRow(data: ValueMatrix[Int, Int, SheetCell], rowNum: Int, height: Option[Int]) {  } /// HINT Fyll ut
  private def createCell(sheetCell: Option[SheetCell], colNum: Int, row: Row) {  } /// HINT Fyll ut
}

object PoiAdapter {
  /**
   * Skriv en arbeidsbok til Excelfil.
   * @param bookName Boknavnet (suffikses med .xlsx)
   * @param workbook Referanse til boken
   * @throws java.io.IOException Verden er ikke perfekt
   */
  def writeToFile(bookName: String, workbook: Workbook) {
    val fileName: String = bookName + ".xlsx"
    val out: FileOutputStream = new FileOutputStream(fileName)
    workbook.write(out)
    out.close
  }
}

